### PR TITLE
Process only files smaller than the threshold for OPTIMIZE

### PR DESCRIPTION
**Pull request by findinpath**
## Description

Fixes https://github.com/trinodb/trino/issues/20088

## Additional context and related issues

Commits are taken out of https://github.com/trinodb/trino/pull/20054

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java [hunk 1]
```diff
@@ -253,12 +251,33 @@ private Stream<DeltaLakeSplit> getSplits(
...
     }
 
-    private Set<Map<String, Optional<String>>> findPartitionsWithAtMostOneFile(List<AddFileEntry> addFileEntries)
+    private static Stream<AddFileEntry> filterValidDataFilesForOptimize(List<AddFileEntry> validDataFiles, long maxScannedFileSizeInBytes)
     {
-        return addFileEntries.stream().collect(groupingBy(AddFileEntry::getCanonicalPartitionValues, counting())).entrySet().stream()
-                .filter(entry -> entry.getValue() <= 1)
-                .map(Map.Entry::getKey)
-                .collect(toImmutableSet());
+        Map<Map<String, Optional<String>>, Optional<AddFileEntry>> queuedAddFileEntriesMap = new HashMap<>();
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java:256]:
```
// Value being present is a pending file for given partition.
// Value being absent is a tombstone, indicating files for the partition should be processed.        
```

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java [hunk 2]
```diff
@@ -253,12 +251,33 @@ private Stream<DeltaLakeSplit> getSplits(
...
+                        }
+                        List<AddFileEntry> concatenatedAddFileEntries = ImmutableList.<AddFileEntry>builder()
+                                .add(addFileEntry)
+                                .add(alreadyQueuedAddFileEntry.get())
+                                .build();
+                        queuedAddFileEntriesMap.put(canonicalPartitionValues, Optional.empty());
+                        return concatenatedAddFileEntries.stream();
+                    }
+
+                    if (addFileEntry.getSize() < maxScannedFileSizeInBytes) {
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java]: Before the "Rewrite data file filtering logic for OPTIMIZE statements" change, `findPartitionsWithAtMostOneFile` didn't take `maxScannedFileSizeInBytes` into account.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java [hunk 3]
```diff
@@ -177,7 +176,11 @@ private Stream<DeltaLakeSplit> getSplits(
 
         MetadataEntry metadataEntry = tableHandle.getMetadataEntry();
         boolean isOptimize = tableHandle.isOptimize();
-        Set<Map<String, Optional<String>>> partitionsWithAtMostOneFile = isOptimize ? findPartitionsWithAtMostOneFile(validDataFiles) : ImmutableSet.of();
+        if (isOptimize) {
+            checkArgument(maxScannedFileSizeInBytes.isPresent(), "maxScannedFileSizeInBytes must be provided when performing OPTIMIZE");
+            validDataFiles = filterValidDataFilesForOptimize(validDataFiles, maxScannedFileSizeInBytes.get())
+                    .collect(toImmutableList());
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java:182]: The change isn't an improvement on it's own. The code is just harder to read.
This is a prep for streaming, so maybe i keep reviewing this in  https://github.com/trinodb/trino/pull/20054 as a whole

**Review by findepi — COMMENTED** [review]
"Rewrite data file filtering logic for OPTIMIZE statements"

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java (hunk 2 above)
Comment (findinpath) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeSplitManager.java]: squashed the commits into one - the rewriting commit did actually contain changes in behavior.

**Review by findepi — APPROVED** [review]

**Comment by jkylling** [comment]
Thanks a lot for fixing!

**Pull request merged**